instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+The integration with sklearn supports the scikit-learn compatible libraries,
+it can be enabled by using ``SklearnInstrumentor``.
+
+.. sklearn: https://github.com/scikit-learn/scikit-learn
+
+Usage
+-----
+
+Package instrumentation example:
+
+.. code-block:: python
+
+    from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+
+    # instrument the sklearn library
+    SklearnInstrumentor().instrument()
+
+    # instrument sklearn and other libraries
+    SklearnInstrumentor(packages=["lightgbm", "xgboost"]).instrument()
+
+
+Model intrumentation example:
+
+.. code-block:: python
+
+    from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+    from sklearn.datasets import load_iris
+    from sklearn.ensemble import RandomForestClassifier
+    from sklearn.model_selection import train_test_split
+    from sklearn.pipeline import Pipeline
+
+    X, y = load_iris(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+    model = Pipeline(
+        [
+            ("class", RandomForestClassifier(n_estimators=10)),
+        ]
+    )
+
+    model.fit(X_train, y_train)
+
+    SklearnInstrumentor().instrument_estimator(model)
+
+"""
+import logging
+import os
+from functools import wraps
+from importlib import import_module
+from inspect import isclass
+from pkgutil import iter_modules
+from typing import Callable, Dict, List, MutableMapping, Sequence, Type, Union
+
+from sklearn.base import BaseEstimator
+from sklearn.pipeline import FeatureUnion, Pipeline
+from sklearn.tree import BaseDecisionTree
+from sklearn.utils.metaestimators import _IffHasAttrDescriptor
+
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.sklearn.version import __version__
+from opentelemetry.trace import get_tracer
+from opentelemetry.util.types import Attributes
+
+logger = logging.getLogger(__name__)
+
+
+def implement_spans(
+    func: Callable,
+    estimator: Union[BaseEstimator, Type[BaseEstimator]],
+    attributes: Attributes = None,
+):
+    """Wrap the method call with a span.
+
+    Args:
+        func: A callable to be wrapped in a span
+        estimator: An instance or class of an estimator
+        attributes: Attributes to apply to the span
+
+    Returns:
+        The passed function wrapped in a span.
+    """
+    if isclass(estimator):
+        name = estimator.__name__
+    else:
+        name = estimator.__class__.__name__
+    logger.debug("Instrumenting: %s.%s", name, func.__name__)
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        with get_tracer(__name__, __version__).start_as_current_span(
+            name="{cls}.{func}".format(cls=name, func=func.__name__),
+            attributes=attributes,
```

Comment:
We shouldn't populate span attributes if span is not recording. See [this](https://github.com/open-telemetry/opentelemetry-python/blob/master/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L140).

Reply:
Is this more appropriate?

```python
    @wraps(func)
    def wrapper(*args, **kwargs):
        with get_tracer(__name__, __version__).start_as_current_span(
            name="{cls}.{func}".format(cls=name, func=func.__name__),
            attributes=attributes,
        ) as span:
            if span.is_recording():
                for k, v in attributes.items():
                    span.set_attribute(k, v)
            return func(*args, **kwargs)

```

Reply:
Yes that's good. Just remove the `attributes=attributes,` in the `start_as_current_span`.